MapboxNavigation/NavigationMapView.swift

```diff
@@ -671,6 +676,188 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
     }
 
     // MARK: - Vanishing route line methods
+
+    private var primaryRoutePoints: RoutePoints?
+    private var primaryRouteLineGranularDistances: RouteLineGranularDistances?
+    private var primaryRouteRemainingDistancesIndex: Int?
+    public var newFractionTraveled: Double = 0.0
+    
+    private struct RoutePoints {
+        var nestedList: [[[CLLocationCoordinate2D]]]
+        var flatList: [CLLocationCoordinate2D]
+    }
+    
+    private struct RouteLineGranularDistances {
+        var distance: Double
+        var distanceArray: [RouteLineDistancesIndex]
+    }
+    
+    private struct RouteLineDistancesIndex {
+        var point: CLLocationCoordinate2D
+        var distanceRemaining: Double
+    }
+    
+    private func initPrimaryRoutePoints(route: Route) {
+        primaryRoutePoints = parseRoutePoints(route: route)
+        primaryRouteLineGranularDistances = calculateRouteGranularDistances(coordinates: primaryRoutePoints?.flatList ?? [])
+    }
+    
+    /**
+    Tranform the route data into nested arrays of legs -> steps -> coordinates.
+    The first and last point of adjacent steps overlap and are duplicated.
+    */
+    private func parseRoutePoints(route: Route) -> RoutePoints {
+        let nestedList = route.legs.map { (routeLeg: RouteLeg) -> [[CLLocationCoordinate2D]] in
+            return routeLeg.steps.map { (routeStep: RouteStep) -> [CLLocationCoordinate2D] in
+                if let routeShape = routeStep.shape {
+                    if !routeShape.coordinates.isEmpty {
+                        return routeShape.coordinates
+                    } else { return [] }
+                } else {
+                    return []
+                }
+            }
+        }
+        let flatList = nestedList.flatMap { $0.flatMap { $0.flatMap { $0 } } }
+        return RoutePoints(nestedList: nestedList, flatList: flatList)
+    }
+    
+    /**
+    Find and cache the index of the upcoming [RouteLineDistancesIndex].
+    */
+    public func updateUpcomingRoutePointIndex(routeProgress: RouteProgress) {
+        let currentLegProgress = routeProgress.currentLegProgress
+        let currentStepProgress = routeProgress.currentLegProgress.currentStepProgress
+        if currentLegProgress != nil && currentStepProgress != nil && primaryRoutePoints != nil {
```

Comment:
Since these properties are not optionals there is no need in additional check.

Reply:
Thanks for the suggestion! I'll remove the unuseful code.